src/shaders/line.fragment.glsl

```diff
@@ -52,9 +53,24 @@ void main() {
 #endif
 
 #ifdef RENDER_LINE_GRADIENT
-    // For gradient lines, v_lineprogress is the ratio along the
-    // entire line, the gradient ramp is stored in a texture.
-    vec4 out_color = texture2D(u_gradient_image, v_uv);
+    // For gradient lines, v_uv.xy are the coord specify where the texture will be simpled.
+    // v_uv[2] and v_uv[3] are specifying the original clip range that the vertex is located in.
+    vec4 out_color = texture2D(u_gradient_image, v_uv.xy);
+    float start = v_uv[2];
+    float end = v_uv[3];
+    float trim_start = u_trim_offset[0];
+    float trim_end = u_trim_offset[1];
+    // v_uv.x is the relative prorgress based on each clip. Calculate the absolute progress based on
+    // the whole line by combining the clip start and end value.
+    float line_progress = (start + (v_uv.x) * (end - start));
+    // Mark the pixel to be transparent when:
+    // 1. trim_offset range is valid
+    // 2. line_progress is within trim_offset range
+    // 3. or if trim_offset end is line end (1.0), mark line_progress >=1.0 part to be transparent to cover 'round'/'square' cap
+    if (trim_end > trim_start && ((line_progress <= trim_end && line_progress >= trim_start) ||
```

Comment:
can this validation be moved from the fragment shader to where it is set as a uniform? and maybe the `trim_end == 1.0` special case can be avoided here by setting it to something > 1 outside the shaders

Reply:
> can this validation be moved from the fragment shader to where it is set as a uniform? 

the problem here is if we move the calculation out of shader, we need to compare the trim offset with the every `line_progress` value, which is generated when creating the vertex array in line_bucket, but the uniform of trim offset is set inside draw_line for each layer, in that sense we need to somehow extract the corresponding line_progress from line bucket and do the comparison, which I think will bring more complexity than the effort of changing shader. 

> and maybe the trim_end == 1.0 special case can be avoided here by setting it to something > 1 outside the shaders

Good idea, I will do that!

Reply:
sorry, by validation I meant just this part `trim_end > trim_start`.

Could it be just `if (trim_start <= line_progress && line_progress <= trim_end) {`